### PR TITLE
Update Civic loader for API v2

### PR DESCRIPTION
**Pull request by mathieulemieux**
This PR updates the CIViC loader to work with their new GraphQL API.

When validating the returned Json file, almost all the fields in the specs needs to be nullable even if not always specified in the official documentation.

Non-nullable fields of evidenceItem observed containing some null values :
- disease
- disease.doid
- drugInteractionType
- drugs.ncitId
- evidenceRating
- source.ascoAbstractId
- source.publicationYear

### CODE REVIEW

File: src/civic/index.js [hunk 1]
```diff
@@ -17,31 +18,31 @@ const { logger } = require('../logging');
...
+const BASE_URL = 'https://civicdb.org/api/graphql';
 
 /**
- * https://civicdb.org/glossary
+ * https://docs.civicdb.org/en/latest/model/evidence/level.html
  */
 const VOCAB = {
-    1: 'Evidence likely does not belong in CIViC. Claim is not supported well by experimental evidence. Results are not reproducible, or have very small sample size. No follow-up is done to validate novel claims.',
-    2: 'Evidence is not well supported by experimental data, and little follow-up data is available. Publication is from a journal with low academic impact. Experiments may lack proper controls, have small sample size, or are not statistically convincing.',
+    1: 'Strong, well supported evidence from a lab or journal with respected academic standing. Experiments are well controlled, and results are clean and reproducible across multiple replicates. Evidence confirmed using independent methods. The study is statistically well powered.',
```
Comment (cgrisdale) [inline src/civic/index.js:35]: Evidence rating of 1 is defined as: 

> Claim is not supported well by experimental evidence. Results are not reproducible, or have very small sample size. No follow-up is done to validate novel claims.

File: src/civic/index.js [hunk 2]
```diff
@@ -17,31 +18,31 @@ const { logger } = require('../logging');
...
+    2: 'Strong, well supported evidence. Experiments are well controlled, and results are convincing. Any discrepancies from expected results are well-explained and not concerning.',
```
Comment (cgrisdale) [inline src/civic/index.js:36]: Evidence rating of 2 is defined as:

> Evidence is not well supported by experimental data, and little follow-up data is available. Publication is from a journal with low academic impact. Experiments may lack proper controls, have small sample size, or are not statistically convincing.

File: src/civic/index.js [hunk 3]
```diff
@@ -17,31 +18,31 @@ const { logger } = require('../logging');
...
 class NotImplementedError extends ErrorMixin {}
 
 const ajv = new Ajv();
 
-const BASE_URL = 'https://civicdb.org/api';
+const BASE_URL = 'https://civicdb.org/api/graphql';
 
 /**
- * https://civicdb.org/glossary
+ * https://docs.civicdb.org/en/latest/model/evidence/level.html
```
Comment (cgrisdale) [inline src/civic/index.js:32]: I think this should be:
`https://docs.civicdb.org/en/latest/model/evidence/evidence_rating.html`

File: src/civic/index.js (hunk 1 above)
Comment (mathieulemieux) [reply src/civic/index.js:35]: Good catch! A copy-paste error from me.

File: src/civic/index.js (hunk 2 above)
Comment (mathieulemieux) [reply src/civic/index.js:36]: Good catch! A copy-paste error from me.

File: src/civic/index.js (hunk 3 above)
Comment (mathieulemieux) [reply src/civic/index.js:32]: There is https://docs.civicdb.org/en/latest/model/evidence/level.html for the evidence Level and https://docs.civicdb.org/en/latest/model/evidence/evidence_rating.html for the Evidence Rating. Do we want both? And what about the url at line 45?

Comment (cgrisdale) [reply src/civic/index.js:32]: I suppose we should have both because we're defining the ratings (1-5) and the levels (A-E), or we could use a more general one for the evidence item page where the other two links can be found:
https://docs.civicdb.org/en/latest/model/evidence.html

Comment (creisle) [reply src/civic/index.js:32]: In the previous version I concatenated them

File: src/civic/index.js [hunk 4]
```diff
@@ -347,10 +381,10 @@ const processEvidenceRecord = async (opt) => {
     const content = {
         conditions: [...variants.map(v => rid(v))],
         description: rawRecord.description,
-        evidence: [rid(publication)],
+        evidence: [rid(publication, true)],
```
Comment (creisle) [inline src/civic/index.js:384]: The publication should never be null. If it is then something has gone wrong and we should throw an error rather than create the statement

File: src/civic/index.js [hunk 5]
```diff
@@ -462,100 +496,78 @@ const processEvidenceRecord = async (opt) => {
  * Get a list of CIViC Evidence Items which have since been deleted.
  * Returns the list of evidence item IDs to be purged from GraphKB
  *
- * @param {string} baseUrl base url for the CIViC API
```
Comment (creisle) [inline src/civic/index.js:465]: why did you remove the variable name here?

File: src/civic/index.js [hunk 6]
```diff
@@ -462,100 +496,78 @@ const processEvidenceRecord = async (opt) => {
...
+                variables: {
+                    status: 'SUBMITTED',
+                    userId: parseInt(curator, 10),
+                },
+            });
+            evidenceItems.push(...submittedByATrustedCurator);
         }
     }
-    logger.info(`loaded ${submitted} unaccepted entries from trusted submitters`);
+    const nbUnaccepted = evidenceItems.length - accepted.length;
```
Comment (creisle) [inline src/civic/index.js:566]: minor: nb isn't a abbreviation i've seen for number before. I would use something more common like num? or even count? `submittedCount` `numSubmitted` etc

File: src/civic/index.js [hunk 7]
```diff
@@ -590,34 +602,38 @@ const downloadEvidenceRecords = async (baseUrl, trustedCurators) => {
 const upload = async ({
     conn, errorLogPrefix, trustedCurators, ignoreCache = false, maxRecords, url = BASE_URL,
 }) => {
-    // add the source node
+    // 1. ADDING CIVIC SOURCE TO GRAPHKB
     const source = await conn.addSource(SOURCE_DEFN);
 
+    // 2. FETCHING PREVIOUS RECORDS FROM GRAPHKB
```
Comment (creisle) [inline src/civic/index.js:608]: are the numbered comments leftover from debugging? They seem superfluous with the other inline comments

File: src/civic/index.js [hunk 8]
```diff
@@ -630,13 +646,15 @@ const upload = async ({
...
+    // Main loop on recordsById
     for (const [sourceId, recordList] of Object.entries(recordsById)) {
         if (previouslyEntered.has(sourceId) && !ignoreCache) {
             counts.exists++;
             continue;
         }
         if (purgeableEvidenceItems.has(sourceId)) {
-            // this should never happen, but if it does we have made an invalida assumption about how civic uses IDs
+            // this should never happen.
+            // If it does we have made an invalid assumption about how civic uses IDs.
```
Comment (creisle) [inline src/civic/index.js:657]: minor: generally I try to avoid multiple inline comments in a row. For this one you don't need multiple lines b/c comments, urls, and strings should be allowed to violate the line length limit. we can just wrap in the editor

File: src/civic/index.js [hunk 9]
```diff
@@ -649,64 +667,67 @@ const upload = async ({
         let mappedCount = 0;
         const postupload = [];
 
-        // resolve combinations
+        // Nested loop on recordList. Resolve combinations
```
Comment (creisle) [inline src/civic/index.js:670]: Try to keep comments more about what it is doing and why vs how it is implemented (unless it is particularly confusing). In this case it should be obvious to the dev this is nested just from the indenting so I would remove this. You could expand it to say something more useful perhaps like "Resolve combinations: splits civic evidence items drugs into separate evidence items based on their combination type. Splits variants into a list to indicate separate evidence items when variants have been linked as "or"

File: src/civic/index.js [hunk 10]
```diff
@@ -649,64 +667,67 @@ const upload = async ({
...
 
             let orCombination;
 
-            if (orCombination = /^([a-z]\d+)([a-z])\/([a-z])$/i.exec(record.variants[0].name)) {
+            if (orCombination = /^([a-z]\d+)([a-z])\/([a-z])$/i.exec(record.variant.name)) {
                 const [, prefix, tail1, tail2] = orCombination;
-                record.variants = [
-                    { ...record.variants[0], name: `${prefix}${tail1}` },
-                    { ...record.variants[0], name: `${prefix}${tail2}` },
+                record.variant = [
```
Comment (creisle) [inline src/civic/index.js:692]: This is incorrect, variants and variant are separate entities. We need the nested loop in the next part. this is important when for example we have a civic evidence item with a variant: "K12G or K12M". In graphkb this must be split into 2 different statements since we do not allow OR combinations. This is why it is stored on "variants" and not "variant" and looped over in the code below

**Review by creisle — CHANGES_REQUESTED** [review]

File: src/civic/index.js (hunk 3 above)
Comment (mathieulemieux) [reply src/civic/index.js:32]: ok. I propose to put them both in the comment section and use https://docs.civicdb.org/en/latest/model/evidence.html for the url field.

File: src/civic/index.js [hunk 11]
```diff
@@ -50,74 +51,99 @@ const RELEVANCE_CACHE = {};
 const validateEvidenceSpec = ajv.compile(evidenceSpec);
 
 
+/**
+ * Request the CIViC GraphQL API
```
[the remark this answers is not in the document]
Comment (mathieulemieux) [reply src/civic/index.js:55]: ok, done.

File: src/civic/index.js (hunk 4 above)
Comment (mathieulemieux) [reply src/civic/index.js:384]: Yeah, we got some errors appening without this argument. I used it because it was already available on the function. I'll roll that back and we'll see exactly which case poses problem.

File: src/civic/index.js (hunk 5 above)
Comment (mathieulemieux) [reply src/civic/index.js:465]: Has been changed by '@param {string} url endpoint for the CIViC API' on line 499.

File: src/civic/index.js [hunk 12]
```diff
@@ -50,74 +51,99 @@ const RELEVANCE_CACHE = {};
...
-    } else if (evidenceDirection === 'Supports') {
+    } else if (evidenceDirection === 'SUPPORTS') {
         switch (evidenceType) { // eslint-disable-line default-case
-            case 'Predictive': {
+            case 'PREDICTIVE': {
                 switch (clinicalSignificance) { // eslint-disable-line default-case
-                    case 'Sensitivity':
-                    case 'Adverse Response':
-                    case 'Reduced Sensitivity':
+                    // case 'Sensitivity':   <-- Deprecated ?
```
[the remark this answers is not in the document]
Comment (mathieulemieux) [reply src/civic/index.js:99]: ok, done.

File: src/civic/index.js [hunk 13]
```diff
@@ -269,7 +301,7 @@ const getEvidenceLevel = async ({
  * @param {object} opt.rawRecord the unparsed record from CIViC
  * @param {object} opt.sources the sources by name
  * @param {boolean} opt.oneToOne civic statements to graphkb statements is a 1 to 1 mapping
- * @param {object} opt.variantsCache keeps track of errors and results processing variants to avoid repeating
+ * @param {object} opt.variantsCache tracks errors and result processing variants to avoid repeating
```
[the remark this answers is not in the document]
Comment (mathieulemieux) [reply src/civic/index.js:304]: ok, done. The previous change was only made in order to keep the line under 100 characters. Not that important after all.

File: src/civic/index.js (hunk 6 above)
Comment (mathieulemieux) [reply src/civic/index.js:566]: It's my french that shines through, lol. submittedCount it is now.

File: src/civic/index.js (hunk 7 above)
Comment (mathieulemieux) [reply src/civic/index.js:608]: It is how I learned to code but I took them out for uniformity, np.

File: src/civic/index.js (hunk 8 above)
Comment (mathieulemieux) [reply src/civic/index.js:657]: ok good

File: src/civic/index.js (hunk 5 above)
Comment (creisle) [reply src/civic/index.js:465]: Ah, yup. This was my mistake. I misread the diff here

File: src/civic/index.js (hunk 9 above)
Comment (mathieulemieux) [reply src/civic/index.js:670]: ok. I put both comments above their respective code section.

File: src/civic/variant.js [hunk 14]
```diff
@@ -297,10 +282,10 @@ const uploadNormalizedVariant = async (conn, normalizedVariant, feature) => {
                 [reference2] = await _entrezGene.fetchAndLoadBySymbol(conn, normalizedVariant.reference2.name);
             }
         }
-        content.reference1 = rid(reference1);
+        content.reference1 = rid(reference1, true);
```
Comment (creisle) [inline src/civic/variant.js:285]: reference 1 can never be null

File: src/civic/variant.js [hunk 15]
```diff
@@ -297,10 +282,10 @@ const uploadNormalizedVariant = async (conn, normalizedVariant, feature) => {
...
 
         if (reference2) {
-            content.reference2 = rid(reference2);
+            content.reference2 = rid(reference2, true);
```
Comment (creisle) [inline src/civic/variant.js:288]: similarly, if we have a reference2 it should also never be null

File: src/civic/index.js (hunk 10 above)
Comment (mathieulemieux) [reply src/civic/index.js:692]: ok. Rolled back to previous implementation. I thought it was related to an older implementation of variant querrying.

File: src/civic/variant.js (hunk 14 above)
Comment (mathieulemieux) [reply src/civic/variant.js:285]: Same as here: https://github.com/bcgsc/pori_graphkb_loader/pull/97#discussion_r862027792
Lets roll back and see what append.

File: src/civic/variant.js (hunk 15 above)
Comment (mathieulemieux) [reply src/civic/variant.js:288]: ok

**Comment by mathieulemieux** [comment]
Here are the stats with the new commit:
{"error":199,"exists":2854,"skip":232,"success":72}

There is also 3x 'TypeError: Cannot read properties of undefined (reading '@rid')'

**Review by creisle — APPROVED** [review]
